CDCgov/prime-reportstream
# Experience/6520 value sets cypress

# Important

This PR depends on https://github.com/CDCgov/prime-reportstream/pull/6719/, please review that first

# Notes

This change introduces an end to end test for the value sets index page, but also introduces two important changes to the cypress testing system.

Cypress runs will now only try to log in through the UI when necessary. Once logged into the UI, cypress will make use of a stored credential from the previous login. This is done to 

1. speed things up and 
2. avoid a situation where an MFA code is being reused too quickly. In a situation where we are logging in on each test, or even each file, we will likely always be trying to reuse an MFA code that has already been used, and Okta doesn't like that

Also, in order to account for lag between:
* okta hook initializing (it returns null on first render even when credentials exist)
* useOktaMemberships hook initializing (it returns null on first render even when an active membership exists)

We have expanded the idea of an `initialized` user state to wait until the membership hook has had a chance to render and return data (we had already made this change to account for the okta hook's initialization in a previous PR https://github.com/CDCgov/prime-reportstream/pull/6811). This change is necessary because cypress runs fast enough that the lag between the okta hook being initialized and the useOktaMemberships hook being initialized is significant.

## TODO:

* - [ ] Fix and update tests

## Test Steps:
1. see https://github.com/CDCgov/prime-reportstream/pull/6719 for directions on running cypress locally

## Changes
- *Include a comprehensive list of changes in this PR*
- *(For web UI changes) Include screenshots/video of changes*

## Checklist

### Testing

### Process

## Linked Issues
- Fixes #issue

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #issue

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?
